doc/design/record_fetcher_proposal.md

```diff
@@ -0,0 +1,170 @@
+to include in our manifest:
+* fetcher metadata
+* what's our entrypoint for this scraper?
+* what commit hash are we running from?
+* a digital signature of the rest of the manifest
+  * this is required to anchor our data lineage
+  * key management is out of scope for this document (but the appendix touches on a half-baked approach)
+* support for extension in the future (we can't be locked in to just these fields)
+
+### Storage Endpoints
+We need to be able to store the raw documents and manifests somewhere durable. We need a specific API to ensure safe
+collection. Our collection API:
+* must allow writing new documents and manifests
+* must not allow for the fetching or altering of existing documents or manifests
+* must enforce authentication and authorization for document submission
+  * this lets us prevent unknown/untrusted entities from aiming a firehose of data at us and eating all of our money,
+  but there's little risk of this threatening the trustworthiness of our records due to the lack of manifest
+  * this document doesn't have a strong opinion here regarding how we authenticate, but the appendix has a strawman we
+can use to start that conversation
+* must not alter the manifest or its signature before committing to durable storage
+  * without the signature, we can't detect tampering or associate the manifest with a fetcher
+* should probably losslessly compress most things (we get 40% savings with gzip on the sample manifest below)
+    
+## Design Ideas
+
+### Document Storage
+First, let's talk document storage. This will be a content digest-addressed blob storage service. We don't need a
+lot of structure for this, as any relationship between documents will be encoded by the manifests -- the service just
+needs to be able to store the raw documents for later lookup (but lookups don't need to be via this service). Backing
+this storage service with a fairly flat filesystem is fine here, we're rarely going to list files, only write and read.
+
+A big benefit of content-addressed storage here is that we get deduping almost for free -- we may need to zero out
+content that changes on every page load (e.g. "This page requested at HH:MM"), but that's more of an optimization than a
+hard requrement. These savings would apply to both documents which remain unchanged between fetches and documents that
+are shared between records.
+
+### Manifest Construction
+
+Manifests need to be signed, so we'll just use [RFC 7519](https://tools.ietf.org/html/rfc7519) JWT blobs, encoding the
+manifest's information in the JWT payload. As an example, the payload could resemble:
+
+```json
+{
+  "version": 1,
+  "uuid": "6e566397-6803-49ec-9280-62c00b368e3f",
+  "record_type": "USA/FL/Bay/Court",
+  "case_id": "20000001",
+  "collected_at_ms": 1592618026000,
+  "fetcher_id": "github.com/Police-Data-Accessibility-Project/Scrapers:USA/FL/Bay/Court/scraper/Scraper.py@1db20ce1e63768363d532d46a641722986f63524",
+  "documents": [
```

Comment:
Hmm. I'm kind of confused, and feel free to let me know if I'm missing something or over complicating things.

It feels like we'd be adding data related to the request in the JWT, If we're adding a signature related to the specific request shouldn't it be in another header, and not connected to the signed jwt?
Since, I think this is essentially a checksum on the data, It's solving a separate problem from auth/auth and and probably should just be included elsewhere in the request.
Maybe we could store each part's checksum in a PDAP_PART_N_DIGEST header or something similar to that? Or a complex object of signatures in a PDAP_DIGEST header.

Reply:
If I'm understanding your concerns correctly, I touch on this in the appendix, but I didn't want to make its suggestions a hard requirement or anything because it's A Lot. The motivation here is that I want every record we fetch to be associable with the code that is running and how it got there, and I want that association to be tamper proof (or at least glaringly tamper evident) because if this takes off, I could see parties interested in tampering with records.

I'm kinda envisioning two JWTs. One that is a bearer token used to authenticate the fetcher, and one that is really a signed encapsulation of our manifest (where only the fetcher holds the private key). The signed manifest format doesn't *have* to be JWT, but it meets the requirements pretty well: signs a JSON payload, flexible crypto options, seems to have good library support in many languages.

As for the digests, while they are checksums, if we use cryptographically strong checksums (say sha256) and persist them to our persistent storage, we can detect tampering of the raw documents. Similarly, by signing all manifests and persisting those signatures to our persistent storage, we can detect tampering at any point in the future by verifying the signature. So by verifying the signature, we can detect tampering in the manifest, which allows us to detect tampering of the underlying documents.

Sorry if this misses what you were going for. Does that help? If so, how could I make this more clear in the document?